event_system.py

```diff
@@ -0,0 +1,48 @@
+from __future__ import annotations
+import json
+import os
+import time
+import multiprocessing as mp
+from dataclasses import dataclass, field, asdict, is_dataclass
+from typing import Any, Callable, Dict, List, Literal, Optional, Protocol, Set, Tuple
+
+class EventWriter:
+    def __init__(self, out_path: str):
+        self.out_path = out_path
+        os.makedirs(os.path.dirname(out_path), exist_ok=True)
+        open(self.out_path, "w").close()
+
+    @staticmethod
+    def _now_iso() -> str:
+        return time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime()) + f".{int((time.time()%1)*1000):03d}Z"
+
+    #ev is the event we want to write
+    def write(self, ev: Dict[str, Any]) -> None:
+        # include timestamp of orchestration
+        if "ts" not in ev:
+            ev["ts"] = self._now_iso()
+        # Convert to JSON-safe format
+        json_safe_ev = self._to_jsonable(ev)
+        with open(self.out_path, "a", encoding="utf-8") as f:
+            f.write(json.dumps(json_safe_ev, ensure_ascii=False) + "\n")
+
+    @staticmethod
+    def _to_jsonable(obj):
+        """Recursively convert to JSON-safe types."""
+        # dataclasses (including slots=True)
+        if is_dataclass(obj):
+            obj = asdict(obj)
```

Comment:
```suggestion
            return asdict(obj)
```
so that it is recursive json instead of an escaped string

Reply:
also not sure about the composition of the dataclasses you pass through here but might be worth just making a dict of the fields and recursively calling your jsonable method on each of the field values